lib2/client/okta.go

```diff
@@ -0,0 +1,667 @@
+	Username string
+	Password string
+	Domain   string
+}
+
+// Checks the validity of OktaCredential and should be called before
+// using the credentials to make API calls.
+//
+// This public method will only validate that credentials exist, it will NOT
+// validate them for correctness. To validate correctness an OktaClient must be
+// used to make a request to Okta.
+func (c *OktaCredential) IsValid() bool {
+	return c.Username != "" && c.Password != "" && c.Domain != ""
+}
+
+// Will fetch your Okta username, password, and domain from your keyring secret
+// backend.
+//
+// Will get the default credentials stored under the `okta-creds` key.
+//
+// feat-request: add support for getting additional sets of credentials.
+// The interface for this functionality needs to be defined, it's
+// possible to then implement alternative credential backends
+// to flexibly support alternative implementations.
+func GetOktaCredentialFromKeyring(kr keyring.Keyring) (OktaCredential, error) {
+	var oktaCreds OktaCredential
+
+	item, err := kr.Get("okta-creds")
+	if err != nil {
+		return oktaCreds, err
+	}
+
+	if err = json.Unmarshal(item.Data, &oktaCreds); err != nil {
+		return oktaCreds, fmt.Errorf("Failed to get okta credentials from your keyring.  Please make sure you have added okta credentials with `aws-okta add`")
+	}
+	return oktaCreds, nil
+}
+
+// looks up the okta domain based on the region. For example, the okta domain
+// for "us" is `okta.com` making your api domain as `<your-org>.okta.com`
+func GetOktaDomain(region string) (string, error) {
+	switch region {
+	case "us":
+		return OktaServerUs, nil
+	case "emea":
+		return OktaServerEmea, nil
+	case "preview":
+		return OktaServerPreview, nil
+	}
+	return "", fmt.Errorf("invalid region %s", region)
+}
+
+// Creates and initializes an OktaClient. This is intended to provide a simple
+// way to create a client that can make requests to the Okta APIs.
+//
+// As an example for how a client might be used:
+// This client can then be passed to a provider that will manage auth
+// for other platforms. Currently AWS SAML provider is supported to get STS
+// credentials to get access to AWS services.
+//
+// Supported configuration options:
+//       TODO: expand on configuration options and add tests.
+//
+// -- proxy config: TBD
+// -- session caching: Passing in a keyring will enable support for caching.
+//      this will cache a valid okta session securely in the keyring. This
+//			session is only for access to the Okta APIs, any additional sessions
+//			(for example, aws STS credentials) will be cached by the provider that
+//      creates them.
+func NewOktaClient(creds OktaCredential, kr *keyring.Keyring, mfaConfig MFAConfig) (*OktaClient, error) {
+
+	if creds.IsValid() {
+		log.Debug("Credentials are valid :", creds.Username, " @ ", creds.Domain)
+	} else {
+		return nil, errors.New("credentials aren't complete. To remedy this, re-add your credentials with `aws-okta add`")
```

Comment:
Good opportunity to use typed errors here. Sketch:

```
type ErrorInvalidCredentials struct {}

...
return nil, &ErrorInvalidCredentials{}
...
```
Then in the CLI code:
```
var errInvalidCreds *client.ErrorInvalidCredentials
if errors.As(err, &errInvalidCredentials) {
  return errors.New("Invalid credentials; try `aws-okta add`")
```
That message is clearly for the CLI user, not the client user, so should be in the CLI.